staging/src/k8s.io/apiserver/pkg/server/long_running_request_terminator.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+type longRunningRequestTerminator struct {
+	tracker map[*context.CancelFunc]struct{}
+	lock    sync.Mutex
+}
+
+func newLongRunningRequestTerminator() *longRunningRequestTerminator {
+	return &longRunningRequestTerminator{
+		tracker: make(map[*context.CancelFunc]struct{}),
+	}
+}
+
+func (l *longRunningRequestTerminator) AddRequest(cancelFn *context.CancelFunc) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	l.tracker[cancelFn] = struct{}{}
+}
+
+func (l *longRunningRequestTerminator) RemoveRequest(cancelFn *context.CancelFunc) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	delete(l.tracker, cancelFn)
+}
+
+func (l *longRunningRequestTerminator) startTerminating(delay time.Duration) {
+	l.lock.Lock()
```

Comment:
Does it cause lock contention here?

it looks like the l.lock will block starting all long running requests.

What if new request starts after startTerminating call?

Reply:
So in most cases after "startTerminating" has been initiated most of requests will never reach longRunningRequestTerminator due to this if `if err := wg.Add(1); err != nil {` in waitgroup.go (sorry, couldn't link to it, no idea why), but there can be a few that will block on this lock till we copy all values from map.

Reply:
Makes sense. thanks!

Reply:
when is the context access concurrently?

Reply:
> when is the context accessed concurrently?
It's not really accessed concurrently, only cancel function is accessed. concurrently.
> can we receive request when we are terminating?
Yes, there is a small chance that request will be received after we start terminating and it won't be canceled.

Reply:
I see, the first lock protect the reads of the map
Then, the goroutine execute the values of the map, that are funcs that cancel and remove themselves (acquiring the lock too) ...

Reply:
Resolving - it seems the comments were addressed.